feat: adds xApiTransforms for completion aggregator events

pomegranited opened the pull request:
**Description:** Describe in a couple of sentence what this PR adds

Defines xAPI event transformers and tests for the aggregator events using the base event-routing-backends transformers.

Based heavily on https://github.com/openedx/event-routing-backends/pull/377 -- thank you @andrey-canon!

Part of https://github.com/openedx/openedx-aspects/issues/222

**JIRA:** [FAL-3766](https://tasks.opencraft.com/browse/FAL-3766) (private link)

**Dependencies:**

* https://github.com/open-craft/openedx-completion-aggregator/pull/206
* https://github.com/openedx/event-routing-backends/pull/431

**Installation instructions:**

1. Install the https://github.com/openedx/tutor-contrib-aspects plugin
1. Add this repo+PR branch to `OPENEDX_EXTRA_PIP_REQUIREMENTS`.
1. Add the repo+PR branch for https://github.com/openedx/event-routing-backends/pull/431 to  `OPENEDX_EXTRA_PIP_REQUIREMENTS` too -- have to do this otherwise Aspects will install its own version.
1. Create a course with sections, and subsections, units, and blocks.

**Testing instructions:**

1. In the LMS: complete some blocks, units, subsections and sections.
2. In Superset: login as a superuser, and visit the SQL tab.
   You should be able to see `progressed` and `completed` events corresponding to `chapter`, `sequential`, `vertical`, and `unit` objects in the `xapi.xapi_events_all_parsed_mv` table.

**Reviewers:**
- [ ] @bmtcril 
- [ ] @Agrendalath 

**Merge checklist:**

- [ ] Revert the TEMP commit and update requirements after https://github.com/openedx/event-routing-backends/pull/431 merges.
- [ ] All reviewers approved
- [x] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

pomegranited commented on completion_aggregator/settings/common.py:
```diff
@@ -32,6 +32,30 @@ def plugin_settings(settings):
         'sequential',
         'vertical',
     }
+
+    # Whitelist the allowed xAPI-transformed events for use with the xAPI backends
+    enabled_aggregator_events = [
+        f'openedx.completion_aggregator.{event_type}.{block_type}'
+
+        for event_type in settings.ALLOWED_COMPLETION_AGGREGATOR_EVENT_TYPES
+        for block_type in settings.ALLOWED_COMPLETION_AGGREGATOR_EVENT_TYPES[event_type]
+    ]
+    global_processors = settings.EVENT_TRACKING_BACKENDS.get(
+        'event_transformer', {}).get(
+            'OPTIONS', {}).get(
+                'processors', [])
+    xapi_processors = settings.EVENT_TRACKING_BACKENDS.get(
+        'event_transformer', {}).get(
+            'OPTIONS', {}).get(
+                'backends', {}).get(
+                    'xapi', {}).get(
+                        'OPTIONS', {}).get(
+                            'processors', [])
+    for processor in global_processors + xapi_processors:
+        if 'whitelist' in processor.get('OPTIONS', {}):
+            for event_name in enabled_aggregator_events:
+                processor['OPTIONS']['whitelist'].append(event_name)
+
```
@bmtcril @Ian2012 I'm trying to add these completion_aggregator events to the whitelists, but a) it isn't working, and b) this is really ugly.

So I'm thinking about creating a function in event-routing-backends for this purpose.

1. Do you have any opinions for how this _should_ work / look? 
1. Is there anything else that should be refreshed/changed when this setting gets updated?

I'll have another bash at this tomorrow.

pomegranited replied:
FYI, I've moved this logic into here, but it's still not always working: https://github.com/openedx/event-routing-backends/pull/431/commits/a415ee72083c20961625b14072a4061931686f11

I think it's breaking because the django app plugins aren't loaded in a consistent order -- this approach requires the ERB app to be loaded first, then the completion aggregator, and I don't think we can rely on that.

bmtcril replied:
Hmm. The apps should be loaded in the order of settings.INSTALLED_APPS, so traditionally you would be able to ensure a dependent app loads correctly by putting it at the bottom of the list. I think that when dealing with Tutor plugins they run in the order of the plugins, so when a Tutor plugin like Aspects is shoving things into INSTALLED_APPS it should be in an idempotent order based on where the Tutor plugin falls.

I would expect ERB to be ready first if the Aspects plugin is higher in the Tutor plugins order than whatever is injecting completion-aggregator into the settings (if being installed via plugin).

pomegranited replied:
> I would expect ERB to be ready first if the Aspects plugin is higher in the Tutor plugins order than whatever is injecting completion-aggregator into the settings (if being installed via plugin).

Yep, I hope so too.. I'm rebuilding my stack now, and we'll see.

> so when a Tutor plugin like Aspects is shoving things into INSTALLED_APPS it should be in an idempotent order based on where the Tutor plugin falls.

Does Aspects shove things into INSTALLED_APPS?  I thought ERB was just a django plugin like this repo, making it just another pip requirement, so who knows what order they're loaded in?

But anyway, these are empirical questions that can be answered by doing :) Hoping I'll have a happy stack today and can test these assumptions.

pomegranited replied:
FYI I rebuilt my tutor dev nightly stack and these changes worked a treat :)

bmtcril commented on test_settings.py:
```diff
@@ -68,6 +68,7 @@ def root(*args):
     'waffle',
     'test_utils.test_app',
     'eventtracking.django.apps.EventTrackingConfig',
+    'event_routing_backends',
```
I was able to get tests working by moving completion_aggregator to the bottom of this list, though you could probably just move ERB above it here.

bmtcril commented on test_settings.py:
```diff
@@ -97,6 +98,8 @@ def root(*args):
 USE_TZ = True
 
 EVENT_TRACKING_ENABLED = True
+EVENT_TRACKING_BACKENDS = {}
```
I also added:

```
EVENT_TRACKING_BACKENDS_ALLOWED_XAPI_EVENTS = []
EVENT_TRACKING_BACKENDS_ALLOWED_CALIPER_EVENTS = []
```

bmtcril commented on completion_aggregator/settings/common.py:
```diff
@@ -52,3 +55,16 @@ def plugin_settings(settings):
+        for block_type in settings.ALLOWED_COMPLETION_AGGREGATOR_EVENT_TYPES[event_type]
+    ]
+    settings.EVENT_TRACKING_BACKENDS_ALLOWED_XAPI_EVENTS += enabled_aggregator_events
+    settings.EVENT_TRACKING_BACKENDS.update(event_tracking_backends_config(
```
Also added `settings` as the first arg here.

bmtcril reviewed (commented):
I am not sure if the changes I've suggested are the correct ones, and I have a suspicion that the version of ERB you're using here is before the settings refactor (but also didn't have time to look) but those changes did get tests to pass for me locally.

Agrendalath commented on completion_aggregator/transformers.py:
```diff
@@ -79,3 +84,119 @@ def transform(self, usage_info, block_structure):  # pylint: disable=unused-argu
+        return Result(
+            completion=self.get_data("data.percent") == 1.0,
+            score={
+                "scaled": self.get_data("data.percent") or 0
```
Why do we call it a "scaled" score?

pomegranited replied:
Huh.. I started to answer you by pointing to the [emitted grading events](https://github.com/openedx/event-routing-backends/blob/980b2aa264261430c5b8dfaeea70c02d44e5e08a/event_routing_backends/processors/xapi/event_transformers/grading_events.py#L63), but since this actually a progress % not a score, this line is actually a misuse of the [`tincan.Result.score` attribute](http://rusticisoftware.github.io/TinCanJS/doc/api/latest/classes/TinCan.Score.html).

So I've updated it to use an Extension instead.

Agrendalath commented on completion_aggregator/transformers.py:
```diff
@@ -79,3 +84,119 @@ def transform(self, usage_info, block_structure):  # pylint: disable=unused-argu
+        Get result for xAPI transformed event.
+        """
+        return Result(
+            completion=self.get_data("data.percent") == 1.0,
```
Can it ever be `True`? The code from https://github.com/open-craft/openedx-completion-aggregator/pull/206 already distinguishes `progress` and `completion` events by checking if the `percent < 1`.

If we want to keep it, would it make sense to call it "completed" instead, since it's a boolean?

pomegranited replied:
Unfortunately we don't have a choice about what these attributes are called -- tincan defines a fixed set of attributes for the [Result class](http://rusticisoftware.github.io/TinCanJS/doc/api/latest/classes/TinCan.Result.html).

But let me know if you'd rather I hard-coded `completion=False` for progress events, if this is confusing.

Agrendalath reviewed (changes requested):
@pomegranited, is there a simple configuration to check the sent events without the tutor plugin? I'm getting the `edx.devstack-palm.lms  | 2024-06-17 15:50:18,680 INFO 3682 [eventtracking.backends.async_routing] [user 3] [ip 172.29.0.1] async_routing.py:41 - [EventEmissionExit] skipping event openedx.completion_aggregator.completion.vertical` every time the event should be sent.

pomegranited commented on requirements/base.in:
```diff
@@ -12,3 +12,4 @@ edx-toggles
 event-tracking
 six
 XBlock[django]
+edx-event-routing-backends  # Provides xAPI transforms for aggregator events
```
We've done some testing with Aspects + ERB and palm/quince, and I don't think there were any compatibility issues discovered there -- @bmtcril would know more.

But since I had to make [some modifications to ERB](https://github.com/openedx/event-routing-backends/pull/431) to support this change, it won't be compatible with prior ERB versions anyway.

pomegranited commented on completion_aggregator/settings/common.py:
```diff
@@ -52,3 +55,17 @@ def plugin_settings(settings):
+        for event_type in settings.ALLOWED_COMPLETION_AGGREGATOR_EVENT_TYPES
+        for block_type in settings.ALLOWED_COMPLETION_AGGREGATOR_EVENT_TYPES[event_type]
+    ]
+    settings.EVENT_TRACKING_BACKENDS_ALLOWED_XAPI_EVENTS += enabled_aggregator_events
```
Ok, this is tricky.. and v. important to get right.

Those settings should be provided by the ERB plugin, which, since it's added as a dependency here, *should* be installed first, and therefore the settings should already have these variables. But more so: if the ERB plugin is loaded after this one, then [ERB will replace these variables with its own settings](https://github.com/openedx/event-routing-backends/pull/431/files#diff-24e49be8ca25acc52569b40114f80c58b30f5025af88fe3a8483da4e6358f6efR90), and we'll lose these events. So the order they load in is critical here.

When you installed this branch on your devstack, did it install my ERB branch from https://github.com/openedx/event-routing-backends/pull/431?

pomegranited replied:
Ok, good plan.

I've made your suggested change on ERB: [c5a89fa](https://github.com/openedx/event-routing-backends/pull/431/commits/c5a89fa26917a1b8394b02db178058088a61e1cf), and here 3ff8331fa7ace76a114276781d9391fd700b536c

pomegranited commented on completion_aggregator/transformers.py:
```diff
@@ -79,3 +84,119 @@ def transform(self, usage_info, block_structure):  # pylint: disable=unused-argu
                 block_structure.set_transformer_block_field(block_key, self, self.AGGREGATORS, aggregators)
+
+
+class BaseAggregatorXApiTransformer(XApiTransformer):
```
Totally, see e694f1683094b43264f1357f82cf03635398dbed

Agrendalath reviewed (approved):
@pomegranited, the changes look good to me. Could you please expand the readme section from #206 to include a note about the xAPI events and update the changelog (I think we should bump the minor version once again).

:+1:

* [x] I tested this: I can see the `xapi_tracking` INFO logs with the `Successfully processed event "openedx.completion_aggregator.progress.{block_type}"` message without any additional configuration on the devstack.
* [x] I read through the code
* [x] I checked for accessibility issues: n/a
* [x] Includes documentation: ⚠️ not yet

Agrendalath commented on requirements/constraints.txt:
```diff
@@ -16,3 +16,7 @@
+
+
+# TODO Temporary constraint to fix issues with using event-routing-backends
+git+https://github.com/open-craft/event-routing-backends.git@jill/fix-no-openedx#egg=edx_event_routing_backends
```
Since the PR is merged now, we can remove the changes from `constraints.txt` and define the min constraint in `base.in`.

pomegranited replied:
Done with [812188c](https://github.com/open-craft/openedx-completion-aggregator/pull/205/commits/812188c27034c3d4b61f1f33627f1008e2faf84a).

pomegranited commented:
@Agrendalath I've fixed the event-routing-backends requirement ([812188c](https://github.com/open-craft/openedx-completion-aggregator/pull/205/commits/812188c27034c3d4b61f1f33627f1008e2faf84a)), bumped the package version to 4.2.0 ([a4858e5](https://github.com/open-craft/openedx-completion-aggregator/pull/205/commits/a4858e58b6a086962a8e251a7519f86c008f06c2)) and added a note about xAPI transforms to the README ([3850789](https://github.com/open-craft/openedx-completion-aggregator/pull/205/commits/38507892a655a8643e577e0e287f3882340ffa52)).

Thank you so much for your thorough reviews! I'll merge this now.
